### DISCUSSION TITLE
Can I give a `CREATE MACRO ... AS TABLE ...` a parameter which is meant to be used as the name of the column produced by the macro?

### DISCUSSION THREAD

**Discussion by bzm3r** [duckdb/duckdb#13676] [Q&A]
Suppose I have the following macro:

```sql
CREATE OR REPLACE MACRO hello_world(col_name, series_start, series_end) AS TABLE (
    SELECT generate_series::VARCHAR AS col_name
    FROM generate_series(series_start, series_end)
  );
CREATE OR REPLACE TABLE tbl AS
FROM hello_world('world', 3, 5);
SHOW TABLES;
```

Execution produces:

```
┌─────────┐
│  name   │
│ varchar │
├─────────┤
│ tbl     │
└─────────┘

┌──────────┐
│ col_name │
│ varchar  │
├──────────┤
│ 3        │
│ 4        │
│ 5        │
└──────────┘
```

But I wanted a column named 'world', yet I got a column named `col_name` (the macro parameter's name): so how do I correctly make sure that the macro parameter's value is used as a column name?

**Comment by Tishj** [comment]
I don't think that would be possible, because the names of the columns have to be known at bind time, and this would involve evaluating an expression which isn't done at bind time.

You can use `FROM hello_world(3, 5) t(world)` instead if you were looking for a workaround

**Comment by jgrg** [comment]
There's a new function `query_table()` coming in v1.1.0: https://youtu.be/xX6qnP2H5wk?t=1146
So similar developments for dynamic SQL creation which would address your particular use case might be coming in the future.

**Comment by cmdlineluser** [comment]
It is also possible in the current version using the json serialize methods.

```sql
CREATE OR REPLACE MACRO hello_world(col_name, series_start, series_end) as table (
   FROM json_execute_serialized_sql(json_serialize_sql(
      format('
         SELECT generate_series::VARCHAR as {}
         FROM generate_series({}, {})
      ', col_name, series_start, series_end)
  ))
)
```

**Reply by jgrg** [reply]
Now `json_serialize_sql() | json_execute_serialized_sql()` _is_ more like an SQL eval! Its restriction seems to be that it only works on SELECT statements and `json_execute_serialized_sql()` will only execute one statement.

**Reply by Maxxen** [reply]
DuckDB v1.1 also has the `query` table function that makes this easier, (no json roundtrip) e.g.
```
SELECT * FROM query('SELECT 1 + 2');
┌─────────┐
│ (1 + 2) │
│  int32  │
├─────────┤
│       3 │
└─────────┘
```
